services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java

```diff
@@ -110,7 +111,18 @@ public boolean isValid(RealmModel realm, UserModel user, CredentialInput credent
             return true;
         } else if (OTPCredentialModel.TOTP.equals(credentialData.getSubType())) {
             TimeBasedOTP validator = new TimeBasedOTP(credentialData.getAlgorithm(), credentialData.getDigits(), credentialData.getPeriod(), policy.getLookAheadWindow());
-            return validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8));
+
+            if (policy.isCodeReusable()) {
+                return validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8));
+            } else {
+                boolean isValid = validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8), secretData.getLastValidInterval());
+                if (isValid) {
```

Comment:
I see one small issue that this does not strictly guarantee that code is used really only once. If there are 2 concurrent callers trying to verify code `123456` it can happen that:
- thread1 (EG. valid user) calls this validation and successfully manage to verify `123456`
- thread2 (EG. potential attacker, who somehow stole the last code of valid user) calls this validation and successfully manage to verify `123456`
- thread1 updates DB. However there was already successful validation of the attacker from thread2 (FAIL)

Not sure how realistic this approach is, but just pointing it :-)

Reply:
Interesting thought! Thanks for that. ATM, the new solution doesn't update DB.